File: packages/birocrat/tests/basic.lua
-- All possible questions we can ask the user
local questions = {
	{
		id = 1,
		type = "simple",
		text = "What is your name, user {id}?",
	},
	{
		id = 2,
		type = "simple",
		text = "How old are you, {name}?",
		default = "30",
	},
	{
		id = 3,
		type = "select",
		text = "What is your favourite type of cuisine?",
		options = { "Indian", "Korean", "Japanese", "Chinese", "Italian" },
	},
	{
		id = 4,
		type = "select",
		text = "What levels of spice can you tolerate?",
		options = { "Mild", "Medium", "Hot", "Very Hot", "Extreme Hot" },
		multiple = true,
	},
}

function Main(state, answer, params)
	-- If this is the first time running the script
	if state == nil and answer == nil then
		-- Return the first question
		local question = questions[1]
		if params.id == nil then
			-- This will fail out the whole form, so there's no need to go back here, hence no need
			-- for a valid state
			return { "error", "No ID parameter provided.", {} }
		end
		question.text = question.text:gsub("{id}", params.id)
		return { "question", question, { question = 1 } }
	end

	if state.question == 1 then
		state.name = answer.text
		state.question = 2
		-- Substitute the user's name in (locally, not globally! pure function!)
		local question = questions[2]
		question.text = question.text:gsub("{name}", state.name)

		return { "question", question, state }
	elseif state.question == 2 then
		state.age = tonumber(answer.text)
		if state.age == nil then
			return { "error", "Please enter a valid number." }
		end
		state.question = 3
		return { "question", questions[3], state }
	elseif state.question == 3 then
		state.favourite_cuisine = answer.selected[1]
		state.question = 4

		if state.favourite_cuisine == "Indian" or state.favourite_cuisine == "Korean" then
			return { "question", questions[4], state }
		else
			return {
				"done",
				{
					name = state.name,
					age = state.age,
					favourite_cuisine = state.favourite_cuisine,
				},
			}
		end
	elseif state.question == 4 then
		state.spice_levels = answer.selected
		return {
			"done",
			{
				name = state.name,
				age = state.age,
				favourite_cuisine = state.favourite_cuisine,
				spice_levels = state.spice_levels,
			},
		}
	end
end

File: packages/birocrat/tests/basic.rs
use std::collections::HashMap;

use birocrat::*;
use mlua::Lua;
use serde_json::json;

static BASIC_SCRIPT: &str = include_str!("basic.lua");

#[test]
fn should_work() {
    let mut params = HashMap::new();
    params.insert("id", 37);
    let vm = Lua::new();
    let mut form = Form::new(BASIC_SCRIPT, params, &vm).unwrap();

    let question = form.first_question();
    assert_eq!(
        question,
        &Question::Simple {
            prompt: "What is your name, user 37?".to_string(),
            default: None,
        }
    );
    let poll = form
        .progress_with_answer(0, Answer::Text("Alice".to_string()))
        .unwrap();
    assert_eq!(
        poll,
        FormPoll::Question {
            question: &Question::Simple {
                prompt: "How old are you, Alice?".to_string(),
                default: Some("30".to_string()),
            },
            answer: None
        }
    );
    // Provide an incorrect answer
    let poll = form
        .progress_with_answer(1, Answer::Text("twenty-five".to_string()))
        .unwrap();
    assert_eq!(
        poll,
        FormPoll::Error("Please enter a valid number.".to_string())
    );
    // Provide a correct answer
    let poll = form
        .progress_with_answer(1, Answer::Text("25".to_string()))
        .unwrap();
    assert_eq!(
        poll,
        FormPoll::Question {
            question: &Question::Select {
                prompt: "What is your favourite type of cuisine?".to_string(),
                default: None,
                options: vec!["Indian", "Korean", "Japanese", "Chinese", "Italian"]
                    .into_iter()
                    .map(|s| s.to_string())
                    .collect(),
                multiple: false
            },
            answer: None
        }
    );
    // Get the previous question and answer
    assert_eq!(
        form.get_question(1),
        Some((
            &Question::Simple {
                prompt: "How old are you, Alice?".to_string(),
                default: Some("30".to_string()),
            },
            Some(&Answer::Text("25".to_string()))
        ))
    );

    // Invalid options
    assert!(form
        .progress_with_answer(2, Answer::Text("Test".to_string()))
        .is_err());
    assert!(form
        .progress_with_answer(
            2,
            Answer::Options(vec!["Indian".to_string(), "Korean".to_string()])
        )
        .is_err());
    assert!(form
        .progress_with_answer(2, Answer::Options(vec!["American".to_string()]))
        .is_err());

    // If we answer with `Italian`, we should be done
    let poll = form
        .progress_with_answer(2, Answer::Options(vec!["Italian".to_string()]))
        .unwrap();
    assert_eq!(poll, FormPoll::Done);

    // But we can go back and answer with something else to get another question
    let poll = form
        .progress_with_answer(2, Answer::Options(vec!["Indian".to_string()]))
        .unwrap();
    assert_eq!(
        poll,
        FormPoll::Question {
            question: &Question::Select {
                prompt: "What levels of spice can you tolerate?".to_string(),
                default: None,
                options: vec!["Mild", "Medium", "Hot", "Very Hot", "Extreme Hot"]
                    .into_iter()
                    .map(|s| s.to_string())
                    .collect(),
                multiple: true,
            },
            answer: None,
        }
    );

    // Now that there's another question, we can't complete the form
    let res = form.into_done();
    assert!(res.is_err());
    form = res.unwrap_err();

    // Answering this with one or more options will work, but not with invalid options
    assert!(form
        .progress_with_answer(
            3,
            Answer::Options(vec!["Mild".to_string(), "Not that hot".to_string()])
        )
        .is_err());
    assert_eq!(
        form.progress_with_answer(3, Answer::Options(vec!["Mild".to_string()]))
            .unwrap(),
        FormPoll::Done
    );
    assert_eq!(
        form.progress_with_answer(
            3,
            Answer::Options(vec!["Mild".to_string(), "Medium".to_string()])
        )
        .unwrap(),
        FormPoll::Done
    );

    // And now we can get the form's final details
    let res = form.into_done().unwrap();
    assert_eq!(
        res,
        json!({
            "name": "Alice",
            "age": 25,
            "favourite_cuisine": "Indian",
            "spice_levels": ["Mild", "Medium"]
        })
    );
}
